Advanced controls for aspect ratio, resolution, and integer scaling.

sheepytina opened the pull request:
Adds new "Advanced Resolution Settings" UI controls and logic to give players greater control over the game image.
- Option to force arbitrary aspect ratios, with some common presets included.
- Option to override Internal Resolution scale with a static vertical pixel count.
- "Pixel-perfect mode" to allow 1:1 pixel mapping and integer scaling.

Aspect ratio and pixel count settings are separate so they can be used independently, and fall back on window size if not being used.

The UI and certain logic is all on the Ship of Harkinian side. Image sizing and positioning is done in Libultraship's Gui functions.

This depends on the following Libultraship pull requests:
- https://github.com/Kenix3/libultraship/pull/335
- https://github.com/Kenix3/libultraship/pull/339

![image](https://github.com/HarbourMasters/Shipwright/assets/99330992/382fc016-a376-4ae4-9f81-71d2cfcca9e5)

https://github.com/HarbourMasters/Shipwright/assets/99330992/ee59b945-a66e-418d-970f-8905aa3bc576

This does not interfere with N64 Mode. N64 Mode (gLowResMode) will take priority over these settings.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890090.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890092.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890094.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890097.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890099.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890100.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1090890101.zip)
<!--- section:artifacts:end -->

sheepytina committed:
Advanced Resolution Settings

first working version with most features implemented

sheepytina committed:
Update advancedResolutionEditor.cpp

Added auto-resizing logic for Pixel Perfect Mode.
Minor fixes.

sheepytina committed:
Merge branch 'develop' of https://github.com/sheepytina/Shipwright into develop

sheepytina committed:
Tweaks and tidying up.

Disable integer scale slider if automatic sizing is overriding it.
Don't offer these UI options on Apple.
Removed unused code.
Updated LUS.

Malkierian commented:
Just making note, requires https://github.com/Kenix3/libultraship/pull/335

sheepytina committed:
Update libultraship

PurpleHato commented on soh/soh/SohGui.cpp:
```diff
@@ -38,6 +38,8 @@
 #include "Enhancements/game-interactor/GameInteractor.h"
 #include "Enhancements/cosmetics/authenticGfxPatches.h"
 
+#include "advancedResolutionEditor.h"
```
Wouldn't it make more sense to place these includes (and so everywhere where they are called) under Enhancements/ResolutionEditor/ instead? We could also include the header and cpp file there. This way, we maintain the workflow for each new type of window being added, keeping things organized similar to other enhancements we've made

sheepytina replied:
Thank you. Does the below commit acceptably resolve this?

PurpleHato commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -253,6 +255,21 @@ void DrawSettingsMenu() {
             };
             UIWidgets::Tooltip("Multiplies your output resolution by the value inputted, as a more intensive but effective form of anti-aliasing");
+           
+            // Tina's resolution thingy
```
I'm not sure if this comment is really needed since it doesn't seem to have a clear purpose. Imagine if everyone started adding their usernames to every section they contributed – it might get a bit cluttered

sheepytina replied:
Yeah I already removed that right after making the PR but hadn't pushed the change yet, don't worry. Will Resolve with the next commit.
(To be clear it was just so i could find my code snippet more easily when i was working, since this is my first time working with this codebase and it's been ages since i've done any coding this serious. I had always intended to remove it.)

PurpleHato reviewed (commented):
Seeing how frequently this feature has been requested, I'm genuinely excited to witness its implementation. Thank you, Tina!

I've noted a couple of small nitpicks I'd like to mention. Apart from those, everything is looking good

sheepytina committed:
Filenames and style fixes

Filenames and folders now more closely match rest of project.
Tidied newlines/comments.
(SohMenuBar.cpp) Label of button changed to fit menu.
(ResolutionEditor.cpp) Default window size improved.

sheepytina committed:
Update libultraship

(However, I still need to make the GUI controls acknowledge the new constraints.)

sheepytina committed:
Update libultraship

(and changed the name of some cvars)

sheepytina committed:
Added constraints to the inputs. Added a fps drop warning.

sheepytina committed:
Tweaks based on feedback

sheepytina committed:
Update libultraship

sheepytina committed:
Enabled on Apple - For currently ongoing Retina DPI troubleshooting.

(Also removed the duplicated N64 Mode toggle.)

sheepytina committed:
Update libultraship

sheepytina committed:
Update LUS, update CVar names, small tweaks

And one significant fix: Enhancement checkboxes in ResolutionEditor now default to off.

sheepytina commented:
Marking this as draft for now as I intend to go back through and polish it up more.

sheepytina commented:
you saw nothing

sheepytina committed:
Add Additional Settings and the accursed horizontal resolution field.

There's still a few bugs with it that I haven't squashed, but I need to stop for now and just commit what I've got.

(This is honestly causing more problems than it solves, but i'm tired of getting questions about it.)

sheepytina committed:
Resolved many of the lingering bugs with the previous commit

sheepytina committed:
Horizontal Resolution field now properly acknowledges resolution bounds.

sheepytina committed:
Don't show "Horiz. pixel count" field if not enforcing aspect ratio.

Additionally:
* Don't change settings if selecting "Custom" from preset dropdowns.
* Added a missing horizontal pixel count clamp check.
* Tidied up redundant behaviour.

sheepytina committed:
Additional comments, and a checkbox to disable aspect correction on consoles.

sheepytina committed:
Change how frame rate threshold is calculated.

sheepytina committed:
More minor UI tweaks.

sheepytina committed:
Added missing CVarSave() calls where needed.

Added a short update countdown for the numerical CVars. This is intended to prevent CVarSave() from being called too often.

sheepytina committed:
Added a helpful button to cover a potential support issue.

sheepytina committed:
"Fit Automatically" has been moved to LUS and is now smarter.

This will require another PR in LUS to be opened by me.

sheepytina committed:
Swap to new branch for libultraship

sheepytina committed:
Even more clever integer scaling behavior.

"IntegerScale" is itself now a CVar group.

sheepytina commented:
I'm now ready for this to be reviewed again. I'm at a point where I feel like it's much more polished now.

sheepytina commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -249,11 +251,30 @@ void DrawSettingsMenu() {
+            ImGui::PushStyleColor(ImGuiCol_Border, ImVec4(0.22f, 0.38f, 0.56f, 1.0f));
+            UIWidgets::Spacer(0);
+            if (mAdvancedResolutionSettingsWindow) {
+                // Tina TODO Change name of gAdvancedResolutionEditorEnabled CVar too.
```
I'll remove this TODO comment in the next commit, but this pertained to [a discussion in the Discord about naming conventions](https://discord.com/channels/808039310850130000/1116005171764400138/1146338166123532298) for "window open" CVars.

If there's no objections, I'll leave "gAdvancedResolutionEditorEnabled" as is for now and let it be addressed in a bespoke PR another time.

sheepytina commented on soh/soh/Enhancements/resolution-editor/ResolutionEditor.cpp:
```diff
@@ -0,0 +1,421 @@
+                }
+                CVarSetInteger("gAdvancedResolution.VerticalPixelCount", verticalPixelCount);
+            }
+            // Delay saving this set of CVars by a predetermined length of time, in frames.
+            updateCountdown = countdownStartingValue;
+        }
+        if (updateCountdown > 0) {
+            updateCountdown--;
+        } else {
+            CVarSave();
+        }
```
This is ostensibly to prevent `CVarSave()` being called too often while typing in the number fields, but I don't know if this is all that necessary or not.

sheepytina commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -249,11 +251,30 @@ void DrawSettingsMenu() {
         if (ImGui::BeginMenu("Graphics")) {
         #ifndef __APPLE__
-            if (UIWidgets::EnhancementSliderFloat("Internal Resolution: %d %%", "##IMul", "gInternalResolution", 0.5f, 2.0f, "", 1.0f, true)) {
+            // Disabled on Apple due to a DPI scaling bug.
+            if (UIWidgets::EnhancementSliderFloat("Internal Resolution: %d %%", "##IMul", "gInternalResolution", 0.5f,
+                                                  2.0f, "", 1.0f, true, true,
+                                                  CVarGetInteger("gAdvancedResolution.VerticalResolutionToggle", 0) &&
+                                                      CVarGetInteger("gAdvancedResolution.Enabled", 0) )) {
                 LUS::Context::GetInstance()->GetWindow()->SetResolutionMultiplier(CVarGetFloat("gInternalResolution", 1));
-            };
+            }
             UIWidgets::Tooltip("Multiplies your output resolution by the value inputted, as a more intensive but effective form of anti-aliasing");
         #endif
```
There was also a discussion on Discord about the Internal Resolution option being disabled on Apple systems due to a bug with high-DPI support that crops up on systems equipped with Retina displays. This affects the Internal Resolution option, but also N64 Mode and my proposed Advanced Resolution Mode.

I originally had the button to open the window not visible on macOS, but made it visible to allow play-testing as part of the discussion. Let me know if I should hide it again.

sheepytina reviewed (commented):
At the stage I'm at now, I consider this more or less feature complete. However I do want to flag up a few small lingering loose ends with or adjacent to this PR.

sheepytina committed:
Tidy up comments.

sheepytina committed:
Fix a typo that prevented `IsDroppingFrames()` from working

(Maybe more than a mere typo, but a typo was involved.)

sheepytina committed:
Remove unused and unnecessary variables.

aMannus commented:
On behalf of author, adding do not merge until https://github.com/Kenix3/libultraship/pull/339 is done and merged. Reviews can start before that however.

sheepytina committed:
Group "Integer Scaling" under its own collapsing header

sheepytina committed:
Changed label for the Enabled advanced settings checkbox.

sheepytina committed:
Update libultraship + Formatting pass on ResolutionEditor.cpp

sheepytina committed:
Add `(Select "Off" to disable.)` help text for the aspect ratio setting and hide UI elements accordingly.

Only show the fields if user chooses Custom.
Padding has been shifted accordingly too.

Also fixed a long standing error with the Y field disappearing when modifying X.

sheepytina committed:
Well I suppose that's no-longer necessary.

sheepytina commented:
My second PR https://github.com/Kenix3/libultraship/pull/339 has been merged now.

sheepytina committed:
Update libultraship with commits from main branch (up to e5df3a9)

sheepytina committed:
Merge branch 'develop' into develop-resolutioneditor

sheepytina commented:
This is no-longer being blocked by libultraship, as all the required changes have been bumped to origin.

sheepytina committed:
Tweak comments.

sheepytina committed:
Save current ImGui Combo items as a console variable

to improve user experience.

sheepytina committed:
Change language of NeverExceedBounds checkbox description

to be more affirmative, so it makes more sense.

Add tooltip for NeverExceedBounds checkbox.
Tweak some comments related to additional settings.

sheepytina committed:
Add list of colours to use with TextColored elements.

sheepytina committed:
Move some UI elements around.

Add an extra MSAA slider to the editor window.

sheepytina committed:
Integer Scaling header is DefaultOpen if player has Pixel Perfect Mode active upon window creation.

+ Amend tooltips.

briaguya0 commented:
removed "do not merge" (looks like that should have happened a month ago https://github.com/HarbourMasters/Shipwright/pull/3130#issuecomment-1758766580)

sheepytina committed:
Fix a minor oversight with default configuration.

Fixes an issue where default aspect ratio settings on a fresh SoH configuration weren't matching the defaults assigned in libultraship.
The default values are now 16:9, matching LUS.
Additionally, the combo box now defaults specifically to the 16:9 preset instead of "Custom".

(Fixing the defaults in LUS to be 4:3 isn't worth a LUS bump, so this slight workaround will do for the sake of this PR.)

sheepytina committed:
Make resolution slider `disabled` condition a variable, for readability.

sheepytina committed:
Small tweak to combo item saving

sheepytina committed:
Use `SCREEN_HEIGHT` and `SCREEN_WIDTH` for constraints

sheepytina committed:
Simplify "Show a horizontal resolution field" logic

by using pixel dimensions as the aspect ratio directly, since now this view hides the aspect ratio setting from the user anyway.

sheepytina committed:
Correct aspect ratio visualiser to be un-inverted

+ actually display it as a ratio.

sheepytina committed:
Remove update flags from combo boxes + remove update countdown

+ remove non-functioning 'IsBoolArrayTrue' function.

(The countdown was an okay idea but I didn't implement it correctly. It's better to just keep it simple.)

sheepytina committed:
Merge remote-tracking branch 'upstream/develop-macready' into develop

briaguya0 commented on soh/soh/Enhancements/resolution-editor/ResolutionEditor.cpp:
```diff
@@ -0,0 +1,453 @@
+
+        UIWidgets::Spacer(0);
+        // Vertical Resolution
+        UIWidgets::PaddedEnhancementCheckbox("Set fixed vertical resolution (disables Resolution slider)",
+                                             "gAdvancedResolution.VerticalResolutionToggle", true, false, false, "",
+                                             UIWidgets::CheckboxGraphics::Cross, false);
```
in my quick testing it wasn't clear that this needed to be checked for the pixel count options to do anything. i tried changing the pixel count preset like 7 times and saw nothing change before realizing i needed to check that box

the first solution that comes to my mind is disabling the options that don't work when the box is unchecked, but if you have a reason to keep them interactable/another way to communicate that they won't do anything in mind that's cool too!

sheepytina replied:
Nah I agree with you here. I'll push a change soon.

sheepytina replied:
I've pushed a change for this with your suggestion.

I took it further and also disable everything if the advanced mode is turned off too.

sheepytina committed:
Code review suggestion: disable UI elements conditionally

 (+ tweaks to code style)

sheepytina committed:
Invisible tweaks to the Integer Scaling-related Additional Settings

This looks like a lot but it's mostly just re-arranging a cluttered area of the code for clarity.
Actual changes to functionality are:
* Help text now "disabled" along with the checkbox.
* The NeverExceedBounds checkbox will now reset the unused ExceedBoundsBy cvar if it's been changed.

sheepytina committed:
Assorted small tweaks to comments and variable declarations.

briaguya0 commented:
loving the disable functionality! noticed one slight oddity with the window exceeded warning

![reset](https://github.com/HarbourMasters/Shipwright/assets/70942617/f0d5bd91-26be-4d34-bb74-9f00c1b2b48b)

sheepytina committed:
Code review suggestion: tweak "Window exceeded" warning condition

briaguya0 reviewed (approved):
:shipit:

sheepytina committed:
Missed a thingy.
